staging/src/k8s.io/client-go/discovery/cached/disk/size_limited_disk_cache.go

```diff
@@ -0,0 +1,239 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package disk
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/hex"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
+	"github.com/peterbourgon/diskv"
+	"k8s.io/klog/v2"
+)
+
+// SizeLimitedDiskCache is an implementation of httpcache.Cache that supplements
+// the in-memory map with persistent storage up to a pre-defined maxSize. After
+// the disk cache size breaches the size limit, the oldest files (by file
+// modification times) are evicted until the size limit is satisfied. This cache
+// is meant to be a substitute for "github.com/gregjones/httpcache/diskcache".
+type SizeLimitedDiskCache struct {
+	maxSize     uint64
+	currentSize uint64
+	fileInfos   []fileInfo
+	diskv       *diskv.Diskv
+}
+
+// Get returns the response corresponding to key if present.
+func (c *SizeLimitedDiskCache) Get(key string) (resp []byte, ok bool) {
+	key = keyToFilename(key)
+	resp, err := c.diskv.Read(key)
+	if err != nil {
+		return []byte{}, false
+	}
+	return resp, true
+}
+
+// Set saves a response to the cache as key, and updates the stored
+// fileInfos and currentSize. This function can cause cache evictions.
+func (c *SizeLimitedDiskCache) Set(key string, resp []byte) {
+	key = keyToFilename(key)
+	size := uint64(len(resp))
+	c.trimDiskCache(size)
```

Comment:
Just thinking out loud. `key` might match an existing file, so `size` is not really what we add to the store (we might replace a file with a smaller file). Should we just add the file to the store and *then* trim what goes over the limit? (edit: also, writing might fail so we would zealously trim).

None of this sounds critical though.

Reply:
If this works the way it appears to read, if someone for some reason just writes the same key over and over again, it'll eventually evict everything in the cache, that doesn't seem good? At least merits a TODO?

Reply:
I don't think this is an issue, since the `appendFileInfo` within this `Set` function removes duplicates before appending. There is a unit test that verifies this.